File: crates/ldw/src-rs/ldw/model/resolved/visitor.rs
// Generated on 2024-11-13T09:33:44.739Z
use super::model::*;

pub trait Visitor {
    #[allow(unused_variables)]
    fn handle_model(self: &mut Self, node: &Box<Model>) {
        self.handle_fqn(&node.name);
        if let Some(value) = &node.parent {
            self.handle_model(&value);
        }
        for (_, value) in &node.definitions {
            self.handle_definition(&value);
        }
    }

    #[allow(unused_variables)]
    fn handle_fqn(self: &mut Self, node: &Fqn) {}

    #[allow(unused_variables)]
    fn handle_definition(self: &mut Self, node: &Box<Definition>) {
        self.handle_type(&node.r#type);
    }

    #[allow(unused_variables)]
    fn handle_type(self: &mut Self, node: &Type) {
        match node {
            Type::VoidType(value) => self.handle_void_type(value),
            Type::PrimitiveType(value) => self.handle_primitive_type(value),
            Type::EnumType(value) => self.handle_enum_type(value),
            Type::TypeWithStructure(value) => self.handle_type_with_structure(value),
            Type::NamedTypeReference(value) => self.handle_named_type_reference(value),
        }
    }

    #[allow(unused_variables)]
    fn handle_void_type(self: &mut Self, node: &Box<VoidType>) {}

    #[allow(unused_variables)]
    fn handle_primitive_type(self: &mut Self, node: &PrimitiveType) {}

    #[allow(unused_variables)]
    fn handle_enum_type(self: &mut Self, node: &Box<EnumType>) {}

    #[allow(unused_variables)]
    fn handle_type_with_structure(self: &mut Self, node: &TypeWithStructure) {
        match node {
            TypeWithStructure::SumType(value) => self.handle_sum_type(value),
            TypeWithStructure::ProductType(value) => self.handle_product_type(value),
            TypeWithStructure::GenericType(value) => self.handle_generic_type(value),
        }
    }

    #[allow(unused_variables)]
    fn handle_sum_type(self: &mut Self, node: &Box<SumType>) {
        for value in &node.members {
            self.handle_type(&value);
        }
    }

    #[allow(unused_variables)]
    fn handle_product_type(self: &mut Self, node: &Box<ProductType>) {
        for value in &node.members {
            self.handle_product_member(&value);
        }
    }

    #[allow(unused_variables)]
    fn handle_product_member(self: &mut Self, node: &Box<ProductMember>) {
        self.handle_type(&node.r#type);
    }

    #[allow(unused_variables)]
    fn handle_generic_type(self: &mut Self, node: &GenericType) {
        match node {
            GenericType::MapType(value) => self.handle_map_type(value),
            GenericType::SetType(value) => self.handle_set_type(value),
            GenericType::SequenceType(value) => self.handle_sequence_type(value),
            GenericType::OptionType(value) => self.handle_option_type(value),
        }
    }

    #[allow(unused_variables)]
    fn handle_map_type(self: &mut Self, node: &Box<MapType>) {
        self.handle_type(&node.key_type);
        self.handle_type(&node.value_type);
    }

    #[allow(unused_variables)]
    fn handle_set_type(self: &mut Self, node: &Box<SetType>) {
        self.handle_type(&node.key_type);
    }

    #[allow(unused_variables)]
    fn handle_sequence_type(self: &mut Self, node: &Box<SequenceType>) {
        self.handle_type(&node.element_type);
    }

    #[allow(unused_variables)]
    fn handle_option_type(self: &mut Self, node: &Box<OptionType>) {
        self.handle_type(&node.r#type);
    }

    #[allow(unused_variables)]
    fn handle_named_type_reference(self: &mut Self, node: &Box<NamedTypeReference>) {
        self.handle_fqn(&node.fqn);
    }
}

File: crates/codegen/ldw/src-ts/languages/ldw/model/resolved/visitor.ts
// Generated on 2024-10-15T17:25:57.996Z
import * as Model from './model';

export class Visitor {
    visitModel(node: Model.Model): void {
        this.visitFqn(node.name);
        if (node.parent != undefined) {
            this.visitModel(node.parent);
        }
        node.definitions.forEach((x) => {
            this.visitDefinition(x);
        });
    }

    visitFqn(node: Model.Fqn): void {
        node.forEach((x) => {});
    }

    visitDefinition(node: Model.Definition): void {
        this.visitType(node.type);
    }

    visitType(node: Model.Type): void {
        switch (node.discriminator) {
            case Model.Discriminator.VoidType:
                this.visitVoidType(node);
                break;
            case Model.Discriminator.PrimitiveType:
                this.visitPrimitiveType(node);
                break;
            case Model.Discriminator.EnumType:
                this.visitEnumType(node);
                break;
            case Model.Discriminator.SumType:
            case Model.Discriminator.ProductType:
            case Model.Discriminator.MapType:
            case Model.Discriminator.SetType:
            case Model.Discriminator.SequenceType:
            case Model.Discriminator.OptionType:
                this.visitTypeWithStructure(node);
                break;
            case Model.Discriminator.NamedTypeReference:
                this.visitNamedTypeReference(node);
                break;
        }
    }

    visitVoidType(node: Model.VoidType): void {}

    visitPrimitiveType(node: Model.PrimitiveType): void {}

    visitEnumType(node: Model.EnumType): void {
        node.members.forEach((x) => {});
    }

    visitTypeWithStructure(node: Model.TypeWithStructure): void {
        switch (node.discriminator) {
            case Model.Discriminator.SumType:
                this.visitSumType(node);
                break;
            case Model.Discriminator.ProductType:
                this.visitProductType(node);
                break;
            case Model.Discriminator.MapType:
            case Model.Discriminator.SetType:
            case Model.Discriminator.SequenceType:
            case Model.Discriminator.OptionType:
                this.visitGenericType(node);
                break;
        }
    }

    visitSumType(node: Model.SumType): void {
        node.members.forEach((x) => {
            this.visitType(x);
        });
    }

    visitProductType(node: Model.ProductType): void {
        node.members.forEach((x) => {
            this.visitProductMember(x);
        });
    }

    visitProductMember(node: Model.ProductMember): void {
        this.visitType(node.type);
    }

    visitGenericType(node: Model.GenericType): void {
        switch (node.discriminator) {
            case Model.Discriminator.MapType:
                this.visitMapType(node);
                break;
            case Model.Discriminator.SetType:
                this.visitSetType(node);
                break;
            case Model.Discriminator.SequenceType:
                this.visitSequenceType(node);
                break;
            case Model.Discriminator.OptionType:
                this.visitOptionType(node);
                break;
        }
    }

    visitMapType(node: Model.MapType): void {
        this.visitType(node.keyType);
        this.visitType(node.valueType);
    }

    visitSetType(node: Model.SetType): void {
        this.visitType(node.keyType);
    }

    visitSequenceType(node: Model.SequenceType): void {
        this.visitType(node.elementType);
    }

    visitOptionType(node: Model.OptionType): void {
        this.visitType(node.type);
    }

    visitNamedTypeReference(node: Model.NamedTypeReference): void {
        this.visitFqn(node.fqn);
    }

    visitTrivia(node: Model.Trivia): void {
        switch (node.discriminator) {
            case Model.Discriminator.LineComment:
                this.visitLineComment(node);
                break;
            case Model.Discriminator.BlockComment:
                this.visitBlockComment(node);
                break;
            case Model.Discriminator.Whitespace:
                this.visitWhitespace(node);
                break;
        }
    }

    visitWhitespace(node: Model.Whitespace): void {}

    visitLineComment(node: Model.LineComment): void {}

    visitBlockComment(node: Model.BlockComment): void {}
}
